test/mpi/io/large_count.c

```diff
@@ -0,0 +1,156 @@
+            printf("Each global variable is of size %d bytes (%.1f MB)\n",
+                   gsize[0] * gsize[1], (float) gsize[0] * gsize[1] / 1048576);
+            printf("Each process writes %zd bytes (%.1f MB, %.1f GB)\n",
+                   buf_len, (float) buf_len / 1048576, (float) buf_len / 1073741824);
+        }
+        printf("rank %3d: gsize=%4d %4d start=%4d %4d count=%4d %4d\n", rank,
+               gsize[0], gsize[1], start[0], start[1], count[0], count[1]);
+    }
+
+
+    /* create 2D subarray datatype for fileview */
+    err = MPI_Type_create_subarray(2, gsize, count, start, MPI_ORDER_C, MPI_BYTE, &subType);
+    ERROR("MPI_Type_create_subarray");
+    err = MPI_Type_commit(&subType);
+    ERROR("MPI_Type_commit");
+
+    /* create a filetype by concatenating NVARS subType */
+    for (i = 0; i < NVARS; i++) {
+        array_of_blocklengths[i] = 1;
+        array_of_displacements[i] = gsize[0] * gsize[1] * i;
+        array_of_types[i] = subType;
+    }
+    err = MPI_Type_create_struct(NVARS, array_of_blocklengths,
+                                 array_of_displacements, array_of_types, &filetype);
+    ERROR("MPI_Type_create_struct");
+    err = MPI_Type_commit(&filetype);
+    ERROR("MPI_Type_commit");
+    err = MPI_Type_free(&subType);
+    ERROR("MPI_Type_free");
+
+    /* Create local buffer datatype: each 2D variable is of size LEN x LEN */
+    gsize[0] = LEN;
+    gsize[1] = LEN;
+    start[0] = 0;
+    start[1] = 0;
+    count[0] = LEN - 1; /* -1 to create holes */
+    count[1] = LEN - 1;
+
+    err = MPI_Type_create_subarray(2, gsize, count, start, MPI_ORDER_C, MPI_BYTE, &subType);
+    ERROR("MPI_Type_create_subarray");
+    err = MPI_Type_commit(&subType);
+    ERROR("MPI_Type_commit");
+
+    /* concatenate NVARS subType into a buftype */
+    for (i = 0; i < NVARS; i++) {
+        array_of_blocklengths[i] = 1;
+        array_of_displacements[i] = LEN * LEN * i;
+        array_of_types[i] = subType;
+    }
+
+    /* create a buftype by concatenating NVARS subTypes */
+    err = MPI_Type_create_struct(NVARS, array_of_blocklengths,
+                                 array_of_displacements, array_of_types, &buftype);
+    ERROR("MPI_Type_create_struct");
+    err = MPI_Type_commit(&buftype);
+    ERROR("MPI_Type_commit");
+    err = MPI_Type_free(&subType);
+    ERROR("MPI_Type_free");
+
+    /* allocate a local buffer */
+    buf_len = (size_t) NVARS *LEN * LEN;
+    buf = (char *) malloc(buf_len);
+    for (i = 0; i < buf_len; i++)
+        buf[i] = (char) rank;
+
+    /* open to create a file */
+    err =
+        MPI_File_open(MPI_COMM_WORLD, filename, MPI_MODE_CREATE | MPI_MODE_WRONLY, MPI_INFO_NULL,
+                      &fh);
+    ERROR("MPI_File_open");
+
+    /* set the file view */
+    err = MPI_File_set_view(fh, 0, MPI_BYTE, filetype, "native", MPI_INFO_NULL);
+    ERROR("MPI_File_set_view");
+
+    /* MPI collective write */
+    err = MPI_File_write_all(fh, buf, 1, buftype, &status);
```

Comment:
As I do not have a push permission to this PR branch, I suggest to add the following few lines after the collective write to test a collective read.
```
    err = MPI_File_seek(fh, 0, MPI_SEEK_SET);
    ERROR("MPI_File_seek");

    /* MPI collective read */
    err = MPI_File_read_all(fh, buf, 1, buftype, &status);
    ERROR("MPI_File_read_all");
```

Reply:
good idea! I just pushed that changed.